common/quotas/clocked_rate_limiter.go

```diff
@@ -137,6 +139,8 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
 	case <-waitExpired:
 		return nil
+	case <-l.recycleCh:
```

Comment:
The reason I didn't cancel the reservation here is because I thought that the reservation we are holding when we receive the `l.recycleCh` message is a reservation that _will_ be successfully used by the waiter, who we are freeing.

Now that you say this, I'm thinking that the ideal thing would be to call `reservation.Cancel` on the reservation that was taken and _not_ used (ie. call `reservation.Cancel` within `RecycleToken` or within `task.finish`). However, those places don't have access to the token.

Since the effect of `reservation.Cancel` is just to restore tokens to the rate limiter, considering that other reservations have already been made, it seems safe to cancel the "successful" reservation as a proxy for cancelling the "failed" reservation that we don't have access to.

TL;DR I've convinced myself that I agree we need to cancel the reservation here